### ISSUE TITLE
@Param("::columns") List<DataTableColumn> 映射的数组顺序和前台传递的不一致

### ISSUE THREAD

**Issue by Wizzercn** [nutzam/nutz#1109]
{"data": "busOrderStatus", "bSortable": true},
                    {"data": "busOrderStatus_1", "bSortable": true},
                    {"data": "busOrderStatus_2", "bSortable": true},

去掉 busOrderStatus 字段就正常

**Comment by wendal** [comment]
我不知道这样的issue能怎么重现.

**Comment by apgzs** [comment]
<%
layout("/layouts/private.html"){
%>
<section class="content-wrap bg-white">
    <header class="header navbar bg-white shadow">
        <div class="btn-group tool-button">
            <a id="add" class="btn btn-primary navbar-btn" href="${base}/private/business/bus_orders/add" data-pjax><i class="ti-plus"></i> 新建 busOrder</a>
            <button class="btn btn-danger navbar-btn" onclick="delCheck()"><i class="ti-close"></i> 删除选中</button>
        </div>
    </header>
    <div class=panel-body style="padding-top: 50px;">
        <div class="table-responsive no-border">
            <input id="unitid" type="hidden">
            <table class="table table-bordered table-striped mg-t datatable">
                <thead>
                    <tr>
                            <th>商品主图</th>
                            <th>营销单号</th>
                            <th>商品名称</th>
                            <th>商品Id</th>
                            <th>商品链接</th>
                            <th>优惠券链接</th>
                            <th>渠道</th>
                            <th>业务员</th>
                            <th>ffffff</th>
                            <th>eeeee</th>
                            <th>ddddd</th>
                            <th>ccccc</th>
                            <th>bbbbb</th>
                            <th>aaaaaa</th>
                            <th>操作</th>
                    </tr>
                </thead>
            </table>
        </div>
    </div>
</section>
<div id="dialogDelete" class="modal fade bs-modal-sm" tabindex="-2" role="dialog" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal" aria-hidden="true">×</button>
                <h4 class="modal-title">删除busOrder</h4>
            </div>
            <div class="modal-body">
                <div class="row">
                    <div class="col-xs-12">
                        busOrder删除后无法恢复。<br/>
                        <br/>确定删除吗？
                    </div>
                </div>
            </div>
            <div class="modal-footer">
                <button type="button" class="btn btn-default" data-dismiss="modal">取  消</button>
                <button id="okDel" type="button" class="btn btn-primary" data-loading-text="正在删除...">确  定</button>
            </div>
        </div>
    </div>
</div>
<div id="dialogDetail" class="modal fade bs-modal-sm" tabindex="-3" role="dialog" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
        </div>
    </div>
</div>
<script language="JavaScript">
    var datatable;
    function initDatatable() {
        datatable = $('.datatable').DataTable({
            "dom": '<"toolbar">frtip',
            "searching":false,
            "processing": false,
            "serverSide": true,
            "select": true,
            "ordering": true,
            "language": {
                "url": "${base}/assets/plugins/datatables/cn.json"
            },
            "preDrawCallback": function () {
                sublime.showLoadingbar($(".main-content"));
            },
            "drawCallback": function () {
                sublime.closeLoadingbar($(".main-content"));
            },
            "ajax": {
                "url": "${base}/private/business/bus_orders/data",
                "type": "post",
                "data": function (d) {
                    d.unitid = $('#unitid').val();
                }
            },
//            "order": [[0, "desc"]],
            "columns": [
                    {"data": "goodsPicMain", "bSortable": true},
                    {"data": "marketingNum", "bSortable": true},
                    {"data": "goodsName", "bSortable": true},
                    {"data": "goodsId", "bSortable": true},
                    {"data": "goodsUrl", "bSortable": true},
                    {"data": "couponUrl", "bSortable": true},
                    {"data": "channel", "bSortable": true},
                    {"data": "salesman", "bSortable": true},
                    {"data": "ffffff", "bSortable": true},
                    {"data": "eeeee", "bSortable": true},
                    {"data": "ddddd", "bSortable": true},
                    {"data": "ccccc", "bSortable": true},
                    {"data": "bbbbb", "bSortable": true},
                    {"data": "aaaaaa", "bSortable": true},
            ],
            "columnDefs": [
                {
                    "render": function (data, type, row) {
                        return '<div class="btn-group"><button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown">' +
                                ' <i class="ti-settings"></i> <span class="ti-angle-down"></span></button><ul class="dropdown-menu" role="menu">' +
                                '<li><a href="${base}/private/business/bus_orders/detail/' + row.id + '" data-toggle="modal" data-target="#dialogDetail">查看信息</a></li>' +
                                '<li><a href="${base}/private/business/bus_orders/edit/' + row.id + '" data-pjax>修改信息</a></li>' +
                                '<li class="divider"></li>' +
                                '<li><a href="javascript:;" onclick="del(\'' + row.id + '\')">删除</a></li>' +
                                '</ul></div>';
                    },
                    "targets": 14
                }
            ]
        });
        datatable.on('click', 'tr', function () {
            $(this).toggleClass('selected');
        });
        $("#searchBtn").on('click', function () {
            datatable.ajax.reload();
        });
    }
    function del(id) {
        var dialog = $("#dialogDelete");
        dialog.modal("show");
        dialog.find("#okDel").unbind("click");
        dialog.find("#okDel").bind("click", function (event) {
            var btn = $(this);
            btn.button("loading");
            $.post("${base}/private/business/bus_orders/delete/" + id, {}, function (data) {
                if (data.code == 0) {
                    datatable.ajax.reload(null,false);
                } else {
                    Toast.error(data.msg);
                }
                //重置按钮状态，关闭提示框
                btn.button("reset");
                dialog.modal("hide");
            }, "json");
        });
    }
    function delCheck() {
        var chks = datatable.rows('.selected').data();
        if (chks.length > 0) {
            var ids = [];
            $.each(datatable.rows('.selected').data(), function (i, n) {
                ids.push(n.id);
            });
            var dialog = $("#dialogDelete");
            dialog.modal("show");
            dialog.find("#okDel").unbind("click");
            dialog.find("#okDel").bind("click", function (event) {
                var btn = $(this);
                btn.button("loading");
                $.post("${base}/private/business/bus_orders/delete", {ids: ids.toString()}, function (data) {
                    if (data.code == 0) {
                        datatable.ajax.reload(null,false);
                    } else {
                        Toast.error(data.msg);
                    }
                    btn.button("reset");
                    dialog.modal("hide");
                }, "json");
            });
        } else {
            Toast.warning("请先选择要删除的busOrder！");
        }
    }
    $(function () {
        initDatatable();
    });
</script>

<%}%>

**Comment by apgzs** [comment]
`<%
layout("/layouts/private.html"){
%>
<section class="content-wrap bg-white">
    <header class="header navbar bg-white shadow">
        <div class="btn-group tool-button">
            <a id="add" class="btn btn-primary navbar-btn" href="${base}/private/business/bus_orders/add" data-pjax><i class="ti-plus"></i> 新建 busOrder</a>
            <button class="btn btn-danger navbar-btn" onclick="delCheck()"><i class="ti-close"></i> 删除选中</button>
        </div>
    </header>
    <div class=panel-body style="padding-top: 50px;">
        <div class="table-responsive no-border">
            <input id="unitid" type="hidden">
            <table class="table table-bordered table-striped mg-t datatable">
                <thead>
                    <tr>
                            <th>商品主图</th>
                            <th>营销单号</th>
                            <th>商品名称</th>
                            <th>商品Id</th>
                            <th>商品链接</th>
                            <th>优惠券链接</th>
                            <th>渠道</th>
                            <th>业务员</th>
                            <th>ffffff</th>
                            <th>eeeee</th>
                            <th>ddddd</th>
                            <th>ccccc</th>
                            <th>bbbbb</th>
                            <th>aaaaaa</th>
                            <th>操作</th>
                    </tr>
                </thead>
            </table>
        </div>
    </div>
</section>
<div id="dialogDelete" class="modal fade bs-modal-sm" tabindex="-2" role="dialog" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header">
                <button type="button" class="close" data-dismiss="modal" aria-hidden="true">×</button>
                <h4 class="modal-title">删除busOrder</h4>
            </div>
            <div class="modal-body">
                <div class="row">
                    <div class="col-xs-12">
                        busOrder删除后无法恢复。<br/>
                        <br/>确定删除吗？
                    </div>
                </div>
            </div>
            <div class="modal-footer">
                <button type="button" class="btn btn-default" data-dismiss="modal">取  消</button>
                <button id="okDel" type="button" class="btn btn-primary" data-loading-text="正在删除...">确  定</button>
            </div>
        </div>
    </div>
</div>
<div id="dialogDetail" class="modal fade bs-modal-sm" tabindex="-3" role="dialog" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
        </div>
    </div>
</div>
<script language="JavaScript">
    var datatable;
    function initDatatable() {
        datatable = $('.datatable').DataTable({
            "dom": '<"toolbar">frtip',
            "searching":false,
            "processing": false,
            "serverSide": true,
            "select": true,
            "ordering": true,
            "language": {
                "url": "${base}/assets/plugins/datatables/cn.json"
            },
            "preDrawCallback": function () {
                sublime.showLoadingbar($(".main-content"));
            },
            "drawCallback": function () {
                sublime.closeLoadingbar($(".main-content"));
            },
            "ajax": {
                "url": "${base}/private/business/bus_orders/data",
                "type": "post",
                "data": function (d) {
                    d.unitid = $('#unitid').val();
                }
            },
//            "order": [[0, "desc"]],
            "columns": [
                    {"data": "goodsPicMain", "bSortable": true},
                    {"data": "marketingNum", "bSortable": true},
                    {"data": "goodsName", "bSortable": true},
                    {"data": "goodsId", "bSortable": true},
                    {"data": "goodsUrl", "bSortable": true},
                    {"data": "couponUrl", "bSortable": true},
                    {"data": "channel", "bSortable": true},
                    {"data": "salesman", "bSortable": true},
                    {"data": "ffffff", "bSortable": true},
                    {"data": "eeeee", "bSortable": true},
                    {"data": "ddddd", "bSortable": true},
                    {"data": "ccccc", "bSortable": true},
                    {"data": "bbbbb", "bSortable": true},
                    {"data": "aaaaaa", "bSortable": true},
            ],
            "columnDefs": [
                {
                    "render": function (data, type, row) {
                        return '<div class="btn-group"><button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown">' +
                                ' <i class="ti-settings"></i> <span class="ti-angle-down"></span></button><ul class="dropdown-menu" role="menu">' +
                                '<li><a href="${base}/private/business/bus_orders/detail/' + row.id + '" data-toggle="modal" data-target="#dialogDetail">查看信息</a></li>' +
                                '<li><a href="${base}/private/business/bus_orders/edit/' + row.id + '" data-pjax>修改信息</a></li>' +
                                '<li class="divider"></li>' +
                                '<li><a href="javascript:;" onclick="del(\'' + row.id + '\')">删除</a></li>' +
                                '</ul></div>';
                    },
                    "targets": 14
                }
            ]
        });
        datatable.on('click', 'tr', function () {
            $(this).toggleClass('selected');
        });
        $("#searchBtn").on('click', function () {
            datatable.ajax.reload();
        });
    }
    function del(id) {
        var dialog = $("#dialogDelete");
        dialog.modal("show");
        dialog.find("#okDel").unbind("click");
        dialog.find("#okDel").bind("click", function (event) {
            var btn = $(this);
            btn.button("loading");
            $.post("${base}/private/business/bus_orders/delete/" + id, {}, function (data) {
                if (data.code == 0) {
                    datatable.ajax.reload(null,false);
                } else {
                    Toast.error(data.msg);
                }
                //重置按钮状态，关闭提示框
                btn.button("reset");
                dialog.modal("hide");
            }, "json");
        });
    }
    function delCheck() {
        var chks = datatable.rows('.selected').data();
        if (chks.length > 0) {
            var ids = [];
            $.each(datatable.rows('.selected').data(), function (i, n) {
                ids.push(n.id);
            });
            var dialog = $("#dialogDelete");
            dialog.modal("show");
            dialog.find("#okDel").unbind("click");
            dialog.find("#okDel").bind("click", function (event) {
                var btn = $(this);
                btn.button("loading");
                $.post("${base}/private/business/bus_orders/delete", {ids: ids.toString()}, function (data) {
                    if (data.code == 0) {
                        datatable.ajax.reload(null,false);
                    } else {
                        Toast.error(data.msg);
                    }
                    btn.button("reset");
                    dialog.modal("hide");
                }, "json");
            });
        } else {
            Toast.warning("请先选择要删除的busOrder！");
        }
    }
    $(function () {
        initDatatable();
    });
</script>

<%}%>`

**Comment by apgzs** [comment]
@Name
    @Column
    @Prev(els = {@EL("uuid()")})
    private String id;

    @Column
    @Comment("商品主图")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String goodsPicMain;
    @Column
    @Comment("营销单号")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String marketingNum;
    @Column
    @Comment("商品名称")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String goodsName;
    @Column
    @Comment("商品Id")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String goodsId;
    @Column
    @Comment("商品链接")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String goodsUrl;
    @Column
    @Comment("优惠券链接")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String couponUrl;
    @Column
    @Comment("渠道")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String channel;
    @Column
    @Comment("业务员")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String salesman;
    @Column
    @Comment("ffffff")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String ffffff;

    @Column
    @Comment("eeeee")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String eeeee;

    @Column
    @Comment("ddddd")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String ddddd;

    @Column
    @Comment("ccccc")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String ccccc;

    @Column
    @Comment("bbbbb")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String bbbbb;

    @Column
    @Comment("aaaaaa")
    @ColDefine(type = ColType.VARCHAR, width = 100)
    private String aaaaaa;

**Comment by ywjno** [comment]
`busOrderStatus` 变成 `busOrderStatus_0` 还能再现么

**Comment by wendal** [comment]
原因在ObjectNaviNode依赖不稳定的原始顺序. 改成Integer排序, 并添加对应的testcase
